src/routes.rs

```diff
@@ -458,5 +473,265 @@ pub mod data {
                 .untuple_one()
                 .and_then(warp_sessions::reply::with_session)
         }
+
+
+        pub fn collection_with_token<S: SessionStore, R: Renderer, T: TokenGenerator, D: Storer>(
+            session_store: S,
+            render_engine: R,
+            token_generator: T,
+            data_store: D,
+        ) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone {
+            warp::any()
+                .and(
+                    warp::path!("data" / String / i64 / "fetch_id" / String / String)
+                        .map(|path, index, fetch_id, token| CollectionWithTokenPathParams { path, index, fetch_id, token }),
+                )
+                .and(warp::query::<WithTokenQueryParams>())
+                .and(warp_sessions::request::with_session(
+                    session_store,
+                    Some(CookieOptions {
+                        cookie_name: "sid",
+                        cookie_value: None,
+                        max_age: Some(60),
+                        domain: None,
+                        path: None,
+                        secure: false,
+                        http_only: true,
+                        same_site: Some(SameSiteCookieOption::Strict),
+                    }),
+                ))
+                .and(warp::any().map(move || token_generator.clone().generate_token().unwrap()))
+                .and(warp::any().map(move || render_engine.clone()))
+                .and(warp::any().map(move || data_store.clone()))
+                .and_then(
+                    move |path_params: CollectionWithTokenPathParams,
+                          query_params: WithTokenQueryParams,
+                          session_with_store: SessionWithStore<S>,
+                          token: String,
+                          render_engine: R,
+                          data_store: D| async move {
+                        let mut template_values = HashMap::new();
+                        match &query_params.css {
```

Comment:
If we rewrite paths to be consistent (i.e. `/data/<path>` and everything else is a query param) we can merge the collection and non-collection route logic into one, with a simple conditional checking to see if fetch_id is set in the query params. Thoughts?

Reply:
donzo